File: othello-client/src/app/entities/board/board.state.ts
import { Board } from "./board";
import { Game } from "../../services/game/game";
import { Disc, DiscType } from "../disc/disc";

export class BoardState {

    game: Game;
    currentStateIndex = -1;
    states: Array<Board>;

    constructor(game: Game) {
        this.game = game;
    }

    public initialize(): void {
        let newBoard = new Board(this.game);
        newBoard.initializeStartingPositions();
        this.states = [];
        this.pushNewState(newBoard);
    }

    public pushNewState(board: Board): void {
        if (this.currentStateIndex === this.states.length - 1) {
            this.states.push(board);
            this.currentStateIndex++;
        } else if (this.currentStateIndex < this.states.length - 1 && this.currentStateIndex >= 0) {
            this.states = this.states.slice(0, this.currentStateIndex + 1);
            this.states.push(board);
            this.currentStateIndex++;
        }
    }

    public getCurrentState(): Board {
        if (this.currentStateIndex < this.states.length && this.currentStateIndex >= 0) {
            return this.states[this.currentStateIndex];
        } else {
            return null;
        }
    }

    public goToPreviousState(): void {
        let board = this.getCurrentState();
        this.game.undoMove()
            .then((updatedGame) => {
                let board = this.getPreviousState();
                this.game.updateScores(board);
                this.game.updatePlayerTurns();
                this.game = updatedGame;
            })
            .catch((error) => {
                board.actionResultMessage = error;
            });
    }

    public setBoardPiece(rowNum: number, colNum: number): void {
        // if the board position is empty
        let board = this.getCurrentState();
        if (board.values[rowNum][colNum] !== null) {
            board.actionResultMessage = Board.EMPTY_LOCATION_RULE;
            return;
        }

        // if the move can outflank an opponent disc
        let outFlankedDiscs = board.getOutflankedDiscs(rowNum, colNum);
        if (outFlankedDiscs.length === 0) {
            board.actionResultMessage = Board.OUTFLANK_RULE;

            return;
        }
        // setting board piece
        console.log('Board before: ');
        console.log(board.values);
        console.log('Player ' + this.game.currentPlayer.discType + ' played: Row = ' + rowNum + ', Col = ' + colNum);
        this.game.playMove(rowNum, colNum)
            .then((updatedGame) => {
                let newBoard = board.getCopy();
                newBoard.values[rowNum][colNum] = new Disc(this.game.currentPlayer.discType, rowNum, colNum);
                newBoard.flipDiscs(newBoard.getSameDiscs(outFlankedDiscs), this.game.currentPlayer.discType);
                this.game.updateScores(newBoard);
                this.game.updatePlayerTurns();
                console.log('Board after: ');
                console.log(newBoard.values);
                this.pushNewState(newBoard);
                this.game = updatedGame;
                this.performNoMoveActions();
            })
            .catch((error) => {
                board.actionResultMessage = error;
            });
    }

    private performNoMoveActions(): void {
        if (!this.getCurrentState().isAnyMovePossible()) {
            let player = this.game.currentPlayer.discType === DiscType.Black ? 'Black' : 'White';
            if (window.confirm('There are no moves possible for Player ' + player + '. Press OK to continue or Cancel to undo.')) {
                this.game.updatePlayerTurns();
            } else {
                this.goToPreviousState();
            }
        }
    }



    private getPreviousState(): Board {
        if (this.currentStateIndex === 0) {
            return this.getCurrentState();
        }
        if (this.currentStateIndex > 0) {
            this.currentStateIndex--;
            return this.getCurrentState();
        } else {
            return null;
        }
    }
}

File: othello-new-server/othello-server/src/main/java/gmu/isa681/project/othelloserver/model/BoardState.java
package gmu.isa681.project.othelloserver.model;

import java.util.ArrayList;

import gmu.isa681.project.othelloserver.entity.GameEntity;
import gmu.isa681.project.othelloserver.entity.game.BoardEntity;

public class BoardState {

	private static final String POSTION_NOT_EMPTY = "No place there!";
	private static final String NO_DISCS_FLIPPED = "None of opponents's disc is flipped";
	private static final String NO_MOVES_POSSIBLE = "There are no moves possible for ";

	private GameEntity game;
	private Board board;

	public BoardState(GameEntity game) {
		super();
		this.game = game;
	}

	public Board setBoardPiece(int rowNum, int colNum) {
		// if the board position is empty
		if (board.getValues().get(rowNum).get(colNum).getDiscType() != null) {
			board.setMoveResultMessage(BoardState.POSTION_NOT_EMPTY);
			return null;
		}

		// if the move can outflank an opponent disc
		ArrayList<Disc> outFlankedDiscs = board.getOutflankedDiscs(rowNum, colNum);
		if (outFlankedDiscs.isEmpty()) {
			board.setMoveResultMessage(BoardState.NO_DISCS_FLIPPED);
			return null;
		}
		// setting board piece
		Board newBoard = board.getCopy();
		newBoard.getValues().get(rowNum).set(colNum, new Disc(this.game.getCurrentTurn(), rowNum, colNum));
		newBoard.flipDiscs(newBoard.getSameDiscs(outFlankedDiscs), this.game.getCurrentTurn());
		this.game.updateScores(newBoard);
		this.game.updatePlayerTurns();

		this.performNoMoveActions(newBoard);

		return newBoard;
	}

	private void performNoMoveActions(Board board) {
		if (!board.isAnyMovePossible()) {
			this.game.setCurrentPlayerHasMoves(false);
		}
	}

	public Board initialize(BoardEntity boardToInitializeWith) {
		this.board = new Board(this.game, null);
		if (null == boardToInitializeWith) {
			board.initializeStartingPositions();
		} else {
			board.loadValuesFromString(boardToInitializeWith.getValues(), board.getValues());
		}
		return board;
	}

	public void goToPreviousState() {
		// this.game.undoMove()
		// .then((updatedGame) => {
		// let board = this.getPreviousState();
		// this.game.updateScores(board);
		// this.game.updatePlayerTurns();
		// this.game = updatedGame;
		// })
		// .catch((error) => {
		// board.actionResultMessage = error;
		// });
	}

	public GameEntity getGame() {
		return game;
	}

	public Board getBoard() {
		return board;
	}

	public void setBoard(Board board) {
		this.board = board;
	}

	public void setGame(GameEntity game) {
		this.game = game;
	}

}
